File: Examples/intel_examples/examples_core_c/transc/source/cimatcopy.c
/*******************************************************************************
* Copyright 2010-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
! Content:
!   mkl_cimatcopy - in-place transposition routine,
!   Example Program Text ( C Interface )
!******************************************************************************/
#include <mkl_trans.h>
#include "common_func.h"

int main(int argc, char *argv[])
{
  size_t n=4, m=6; /* rows, cols of source matrix */
  MKL_Complex8 alpha;
  MKL_Complex8 src[] = {
      { 1.,  2.},  {3.,  4.},   {5.,  6.},  {7.,  8.},  {9., 10.}, {11., 12.},
      {13., 14.}, {15., 16.},  {17., 18.}, {19., 20.}, {21., 22.}, {23., 24.},
      {25., 26.}, {27., 28.},  {29., 30.}, {31., 32.}, {33., 34.}, {35., 36.},
      {37., 38.}, {39., 40.},  {41., 42.}, {43., 44.}, {45., 46.}, {47., 48.}
  };/* source matrix */
  alpha.real = 1.;
  alpha.imag = 0.;

  printf("\nExample of using mkl_cimatcopy transposition\n");
  printf("INPUT DATA:\n");
  printf("Source matrix:\n");
  print_matrix(n, m, 'c', src);
                       
  /* 
  **  Submatrix(3,4) will be transposed and the rest of the matrix will be unchanged    
  */
  mkl_cimatcopy('R'    /* row-major ordering */, 
                'T'    /* matrix will be transposed */, 
                 3     /* rows */, 
                 4     /* cols */, 
                 alpha /* scales the input matrix */, 
                 src   /* source matrix */, 
                 6     /* src_lda */, 
                 6     /* dst_lda */);
  /* New matrix: src = {  
  **   1.,  2., 13., 14., 25., 26.,  7.,  8.,  9., 10., 11., 12.,
  **   3.,  4., 15., 16., 27., 28., 19., 20., 21., 22., 23., 24.,
  **   5.,  6., 17., 18., 29., 30., 31., 32., 33., 34., 35., 36.,
  **   7.,  8., 19., 20., 31., 32., 43., 44., 45., 46., 47., 48.      
  ** }
  */
  printf("OUTPUT DATA:\nDestination matrix - submatrix(3,4) will be transposed and the rest of the matrix will be unchanged:\n");
  print_matrix(n, m, 'c', src);

  /*
  **  Submatrix(2,6) will be conjugated and the rest of the matrix will be unchanged    
  */
  mkl_cimatcopy('R'    /* row-major ordering */, 
                'R'    /* matrix will be conjugated */, 
                 2     /* rows */, 
                 6     /* cols */, 
                 alpha /* scales the input matrix */, 
                 src   /* source matrix */, 
                 6     /* src_lda */, 
                 6     /* dst_lda */);
  /* New matrix: src = { 
  **    1., -2.,  13., -14.,  25., -26.,   7.,  -8.,   9., -10.,  11., -12.,
  **    3., -4.,  15., -16.,  27., -28.,  19.,  -20., 21., -22.,  23., -24.,
  **    5.,  6.,  17.,  18.,  29.,  30.,  31.,   32., 33.,  34.,  35.,  36.,
  **    7.,  8.,  19.,  20.,  31.,  32.,  43.,   44., 45.,  46.,  47.,  48.     
  **  }
  */
  /*  
  **  Destination matrix will replace the source matrix with using dst_lda
  */
  printf("Destination matrix - submatrix(2,6) will be conjugated and the rest of the matrix will be unchanged:\n");
  print_matrix(n, m, 'c', src);

  /*
  **  Submatrix(2,6) will be conjugate transposed and the rest of the matrix will be unchanged    
  */
  mkl_cimatcopy('R'    /* row-major ordering */, 
                'C'    /* matrix will be conjugate transposed */, 
                 2     /* rows */, 
                 3     /* cols */, 
                 alpha /* scales the input matrix */, 
                 src   /* source matrix */, 
                 6     /* src_lda */, 
                 6     /* dst_lda */);
  /* New matrix: src = { 
  **    1.,  2.,   3.,  4., 25., -26.,   7.,  -8.,   9., -10.,  11., -12.,
  **   13.,  14., 15., 16., 27., -28.,  19., -20.,  21., -22.,  23., -24.,
  **   25.,  26., 27., 28., 29.,  30.,  31.,  32.,  33.,  34.,  35.,  36.,
  **    7.,  8.,  19., 20., 31.,  32.,  43.,  44.,  45.,  46.,  47.,  48.     
  **  }
  */
  /* 
  **  Destination matrix will replace the source matrix with using dst_lda
  */
  printf("Destination matrix - submatrix(2,6) will be conjugate transposed and the rest of the matrix will be unchanged:\n");
  print_matrix(n, m, 'c', src);

  return 0;
}

File: Examples/intel_examples/examples_core_f/transf/source/cimatcopy.f90
!===============================================================================
! Copyright 2010-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!      C I M A T C O P Y  Example Program Text
!*******************************************************************************
    program  CIMATCOPY_MAIN
    include 'mkl_trans.fi' 

    integer :: i, j
!   source matrix 
    complex, dimension(3,4) :: A  = RESHAPE(&
                                              &(/(1.0, 1.0), (5.0, 5.0), ( 9.0,  9.0),&  
                                                &(2.0, 2.0), (6.0, 6.0), (10.0, 10.0),& 
                                                &(3.0, 3.0), (7.0, 7.0), (11.0, 11.0),&
                                                &(4.0, 4.0), (8.0, 8.0), (12.0, 12.0) /),SHAPE(A))
    complex :: alpha = (1.0, 0.0)

!   Executable Statements
    print*, ' Example of using mkl_cimatcopy transposition '
    print*, ' INPUT DATA: '
    print*, ' Source matrix: '
    
!   print source matrix           
    do i= 1, 3
          print 102, ( A(i, j), j=1, 4)
    enddo
      
! 
!   Submatrix(3, 2) is transposed     
!
    call MKL_CIMATCOPY('C',    & ! column-major ordering
                      &'T',    & ! AB is transposed
                      & 3,     & ! rows of destination matrix
                      & 2,     & ! cols of destination matrix
                      & alpha, & ! alpha
                      & A,     & ! source matrix
                      & 3,     & ! src_lda
                      & 3 )      ! dst_lda
!   New matrix: A =  
!  ( (1.0, 1.0),  (5.0, 5.0), ( 9.0,  9.0), (4.0, 4.0),
!    (2.0, 2.0),  (6.0, 6.0), (10.0, 10.0), (8.0, 8.0),
!    (9.0, 9.0),  (10.0,10.0),(11.0, 11.0), (12.0, 12.0)
!   )          
    print*, 'OUTPUT DATA: '
    print*, 'Destination matrix : '
    print*, 'submatrix(3,2) is transposed'
      
!   print destination matrix      
    do i= 1, 3
          print 102,  ( A(i, j), j = 1, 4)
    enddo

102 format(9x,10(f6.2,',',f6.2,'i',3x))
        
    stop
    end        
